### PR TITLE
HTCONDOR-1175 ep-direct-attach

### PR DESCRIPTION
**Pull request by JaimeFrey**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_schedd.V6/schedd.cpp [hunk 1]
```diff
@@ -7686,6 +7686,75 @@ Scheduler::negotiate(int command, Stream* s)
...
+	dprintf(D_ALWAYS, "CmdDirectAttach() reading %d slot ads\n", num_ads);
+
+	for (int i = 0; i < num_ads; i++) {
+		std::string slot_name;
+		if (!rsock->get_secret(claim_id) || !getClassAd(rsock, slot_ad)) {
+			dprintf(D_ALWAYS, "CmdDirectAttach() failed to read slot ad %d\n", i);
+			return 0;
+		}
+
+			// TODO allow trusted users to match all jobs
```
Comment (Todd-L-Miller) [inline src/condor_schedd.V6/schedd.cpp:7728]: This would be nice for the AWS annex, but as such, is not a high priority.

**Review by Todd-L-Miller — COMMENTED** [review]
There's no version history or documentation, although that may be intentional since the ticket describes "an initial prototype."

There's no test, either, although it seems like a smoke test should not be hard to implement.

The only comment of any seriousness is the one about `ATTR_USER`.

File: src/condor_schedd.V6/qmgmt.cpp [hunk 2]
```diff
@@ -8894,6 +8901,16 @@ void FindRunnableJob(PROC_ID & jobid, ClassAd* my_match_ad,
 				continue;
 			}
 
+			if (!match_user.empty()) {
+				// TODO Get the owner from the JobQueueJob object.
+				//   Need to be careful about whether UID_DOMAIN is included.
```
[the remark this answers is not in the document]
Comment (Todd-L-Miller) [reply src/condor_schedd.V6/qmgmt.cpp:8906]: We discussed this in-person; the short answer is yes.

File: src/condor_schedd.V6/schedd.cpp (hunk 1 above)
Comment (JaimeFrey) [reply src/condor_schedd.V6/schedd.cpp:7811]: Leaving this as future work.

File: src/condor_schedd.V6/schedd.cpp [hunk 3]
```diff
@@ -7686,6 +7686,75 @@ Scheduler::negotiate(int command, Stream* s)
...
+			//   Could use ATTR_NEGOTIATOR_SCHEDDS_ARE_SUBMITTERS
+		slot_ad.Assign(ATTR_AUTHENTICATED_IDENTITY, slot_user);
+		slot_ad.Assign(ATTR_RESTRICT_TO_AUTHENTICATED_IDENTITY, true);
+
+		slot_ad.LookupString(ATTR_NAME, slot_name);
+
+			// TODO handle pre-claimed slots
```
[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_schedd.V6/schedd.cpp:7818]: An earlier iteration of this work allowed the startd to immediately put its slots into Claimed state (for the schedd) before providing them to the schedd. This was part of Miron's desire to spin up a bespoke pool and start jobs running "as fast as possible". That option isn't present in the current implementation.

File: src/condor_schedd.V6/schedd.cpp [hunk 4]
```diff
@@ -7686,6 +7686,75 @@ Scheduler::negotiate(int command, Stream* s)
...
+	ClassAd cmd_ad;
+	int num_ads = 0;
+	std::string claim_id;
+	ClassAd slot_ad;
+	std::string slot_user;
+	std::string slot_submitter;
+	PROC_ID jobid;
+	jobid.cluster = jobid.proc = -1;
+
+	dprintf(D_ALWAYS, "Got DIRECT_ATTACH from %s\n", rsock->peer_description());
```
[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_schedd.V6/schedd.cpp:7785]: Yeah, D_ALWAYS is a bit chatty for these messages. I'm changing them to D_FULLDEBUG.

**Review by Todd-L-Miller — APPROVED** [review]

**Pull request merged**